hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/SqlBackend.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite;
+
+import com.hazelcast.sql.impl.calcite.parse.QueryParseResult;
+import com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeFactory;
+import com.hazelcast.sql.impl.optimizer.OptimizationTask;
+import com.hazelcast.sql.impl.optimizer.SqlPlan;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.prepare.Prepare;
+import org.apache.calcite.prepare.Prepare.CatalogReader;
+import org.apache.calcite.sql.parser.SqlParserImplFactory;
+import org.apache.calcite.sql.util.SqlVisitor;
+import org.apache.calcite.sql.validate.SqlConformance;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql2rel.SqlRexConvertletTable;
+import org.apache.calcite.sql2rel.SqlToRelConverter;
+import org.apache.calcite.sql2rel.SqlToRelConverter.Config;
+
+/**
+ * Provides various customization points for the optimization engine.
+ */
+public interface SqlBackend {
```

Comment:
Minor comment: I still have a worry that this interface might render insufficient for some commands in the future, which will block Jet from making some changes until the relevant codebase is adjusted in IMDG. 

I do not have any specific examples right now, so this is not a comment to address immediately. We just need to be aware of the alternative approach, where the backend just consumes the `SqlQuery` and produces the abstract result. In this case, the risk of required changes in IMDG will be minimal at the cost of more complicated backend implementation in Jet (possibly with some code duplication).

Hopefully, the current approach will work well enough for us.

Reply:
Right, that would require duplicating fair amount of code. Not mentioning, we (Jet) would need to carefully monitor all the commits that go into sql module to replicate the changes on the Jet side (we want exactly same behavior for IMDG queries as if Jet would not be there). Moreover, the logic would most probably look exactly as it does currently in IMDG...

Regarding unknown unknowns, I was thinking about introducing `JetSqlService.execute(SqlQuery sql)` - the logic would could go like this:
```
if (jetSqlService != null) {
    SqlResult result = jetSqlService.execute(query);
    if (result != null) {
         return result;
    }
}
// otherwise continue with current flow
```
Right now, we would leave it returning `null` but if we really encounter a case which does not fit the current solution we could plugin required logic until we get another IMDG release. WDYT?

Reply:
AFAIU we do not have concrete examples where the current approach doesn't work. So perhaps we may leave it as is for now.

Reply:
Also, we might add a default method on the interface with the more generic signature. Shared code can be located in that method and it might call another, more concrete, method on the same interface. If at some point jet needs more control, it can switch from implementing the more concrete method to implementing the more generic method. Hopefully this comment makes sense :)